dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/TemplateAPIImpl.java

```diff
@@ -164,167 +180,377 @@ private void save(final Template template) throws DotDataException {
+		if(templateLiveVersion!=null){
+			APILocator.getVersionableAPI().removeLive(template.getIdentifier());
+		}
+		templateWorkingVersion.setModDate(new java.util.Date());
+		templateWorkingVersion.setModUser(user.getUserId());
+		// sets deleted to true
+		APILocator.getVersionableAPI().setDeleted(templateWorkingVersion, true);
+		templateFactory.save(templateWorkingVersion);
 	}
 
-	@WrapInTransaction
-	public void unarchive (final Template template) {
 
+	@WrapInTransaction
+	public void unarchive (final Template template, final User user)
+			throws DotDataException, DotSecurityException {
 		Logger.debug(this, ()-> "Doing unarchive of the template: " + template.getIdentifier());
-		if (Try.of(()->template.isArchived()).getOrElseThrow(e -> new RuntimeException(e))) {
 
-			Try.run(()->WebAssetFactory.unArchiveAsset(template))
-					.getOrElseThrow(e -> new RuntimeException(e));
+		//Check if the Template is locked, and if is locked by a diff user that is executing the action
+		if(template.isLocked() && !this.versionableAPI.get().getLockedBy(template).get().equals(user.getUserId())){
+			Logger.error(this, "The Template: " + template.getName() + " can not be unarchived. "
+					+ "Because it is locked");
+			throw new DotStateException("Template can not be unarchive because is locked");
+		}
+
+		//Check Edit Permissions over Template
+		if(!this.permissionAPI.doesUserHavePermission(template, PERMISSION_EDIT, user)){
+			Logger.error(this,"The user: " + user.getUserId() + " does not have Permissions to Edit the Template");
+			throw new DotSecurityException("User does not have Permissions to Edit the Template");
+		}
+		// Check that the template is archived
+		if(!isArchived(template)){
+			Logger.error(this, "The Template: " + template.getName() + " can not be unarchive. "
+					+ "Because it is not archived");
+			throw new DotStateException("Template must be archived before it can be unarchived");
 		}
+		template.setModDate(new java.util.Date());
+		template.setModUser(user.getUserId());
+		APILocator.getVersionableAPI().setDeleted(template, false);
+		templateFactory.save(template);
+	}
+
+	public boolean isArchived(final Template template) throws DotDataException {
+		return APILocator.getVersionableAPI().isDeleted(template);
 	}
 
 	@WrapInTransaction
-	public boolean deleteTemplate(final Template template, final User user, final boolean respectFrontendRoles) {
+	public void deleteTemplate(final Template template, final User user, final boolean respectFrontendRoles)
+			throws DotDataException, DotSecurityException {
 
 		Logger.debug(this, ()-> "Doing delete of the template: " + template.getIdentifier());
 
-		if (Try.of(()->template.isArchived()).getOrElseThrow(e -> new RuntimeException(e))) {
+		//Check if the Template is locked, and if is locked by a diff user that is executing the action
+		if(template.isLocked() && !this.versionableAPI.get().getLockedBy(template).get().equals(user.getUserId())){
```

Comment:
There are common validation used in several method that can be extracted to a method, like: check if the template is locked, does the user have permissions